pallets/loans/src/functions.rs

```diff
@@ -125,68 +125,99 @@ impl<T: Config> Pallet<T> {
-					.iter()
-					.enumerate()
-					.find(|(_, active_loan)| active_loan.loan_id == loan_id)
-				{
-					ensure!(
-						active_loan.total_borrowed == Zero::zero(),
-						Error::<T>::LoanIsActive
-					);
-					active_loans.remove(idx);
-				}
-			}
+		ensure!(
+			interest_rate_per_sec >= One::one(),
+			Error::<T>::LoanValueInvalid
+		);
 
-			// ensure loan_type is valid
-			let now = Self::now();
-			ensure!(loan_type.is_valid(now), Error::<T>::LoanValueInvalid);
+		let active_loan = PricedLoanDetails {
+			loan_id,
+			loan_type,
+			interest_rate_per_sec,
+			origination_date: None,
+			normalized_debt: Zero::zero(),
+			total_borrowed: Zero::zero(),
+			total_repaid: Zero::zero(),
+			write_off_status: WriteOffStatus::None,
+			last_updated: now,
+		};
 
-			// ensure interest_rate_per_sec >= one
-			ensure!(
-				interest_rate_per_sec >= One::one(),
-				Error::<T>::LoanValueInvalid
-			);
+		let mut active_loans = ActiveLoans::<T>::get(pool_id);
+		active_loans
+			.try_push(active_loan)
+			.map_err(|_| Error::<T>::TooManyActiveLoans)?;
+		let count = active_loans.len();
+		ActiveLoans::<T>::insert(pool_id, active_loans);
 
-			let active_loan = PricedLoanDetails {
-				loan_id,
-				loan_type,
-				interest_rate_per_sec,
-				origination_date: None,
-				normalized_debt: Zero::zero(),
-				total_borrowed: Zero::zero(),
-				total_repaid: Zero::zero(),
-				write_off_status: WriteOffStatus::None,
-				last_updated: Self::now(),
-			};
-
-			active_loans
-				.try_push(active_loan)
-				.map_err(|_| Error::<T>::TooManyActiveLoans)?;
-			let count = active_loans.len();
-			ActiveLoans::<T>::insert(pool_id, active_loans);
-
-			// update the loan status
-			loan.status = LoanStatus::Active;
-
-			Ok(count.try_into().unwrap())
-		})
+		Ok(count.try_into().unwrap())
+	}
+
+	pub(crate) fn price_active_loan(
+		pool_id: PoolIdOf<T>,
+		loan_id: T::LoanId,
+		interest_rate_per_sec: T::Rate,
+		loan_type: LoanType<T::Rate, T::Balance>,
+	) -> Result<u32, DispatchError> {
+		let now = Self::now();
+		ensure!(loan_type.is_valid(now), Error::<T>::LoanValueInvalid);
+
+		ensure!(
+			interest_rate_per_sec >= One::one(),
+			Error::<T>::LoanValueInvalid
+		);
+
+		Self::try_mutate_active_loan(
+			pool_id,
+			loan_id,
+			|active_loan| -> Result<(), DispatchError> {
+				let old_debt = T::InterestAccrual::previous_debt(
+					active_loan.interest_rate_per_sec,
+					active_loan.normalized_debt,
+					active_loan.last_updated,
+				)?;
+
+				// calculate old present_value
+				let write_off_groups = PoolWriteOffGroups::<T>::get(pool_id);
+				let old_pv = active_loan
+					.present_value(old_debt, &write_off_groups, active_loan.last_updated)
+					.ok_or(Error::<T>::LoanPresentValueFailed)?;
+
+				// calculate new normalized debt without amount
+				let normalized_debt = T::InterestAccrual::adjust_normalized_debt(
+					interest_rate_per_sec,
+					active_loan.normalized_debt,
+					Adjustment::Increase(Zero::zero()),
```

Comment:
I added here an amount of zero because I don't have an amount itself to reevaluate the normalized debt. I understand that if I'm only pricing it should be zero, but please, confirm this 😄

Reply:
This is a good question :) The interest rate here changes, so I actually think we need the new code from Branan's outstanding PR here: https://github.com/centrifuge/centrifuge-chain/pull/870/files#diff-71eb00873e4a9d42ca27d4e664714802c8331f5b7b648962b9ac892552665f1bR154 As in, this should be:

```
let normalized_debt = T::InterestAccrual::renormalize_debt(
  previous_interest_rate_per_sec, // we'd need to retrieve this this
  interest_rate_per_sec,
  active_loan.normalized_debt,
)?;
```

Reply:
That fits much better!

Reply:
#870 is finally merged, this is unblocked